### PR TITLE
learning and tests

### PR DESCRIPTION
**Pull request by S-ign**
## Tips:

* please read the style guide before submitting your patch:
[docs/style-guide.md](../docs/style-guide.md)

* commit message titles must be in the form:

```topic: Capitalized message with no trailing period```

or:

```topic, topic2: Capitalized message with no trailing period```

* golang code must be formatted according to the standard, please run:

```
make gofmt		# formats the entire project correctly
```

or format a single golang file correctly:

```
gofmt -w yourcode.go
```

* please rebase your patch against current git master:

```
git checkout master
git pull origin master
git checkout your-feature
git rebase master
git push your-remote your-feature
hub pull-request	# or submit with the github web ui
```

* after a patch review, please ping @purpleidea so we know to re-review:

```
# make changes based on reviews...
git add -p		# add new changes
git commit --amend	# combine with existing commit
git push your-remote your-feature -f
# now ping @purpleidea in the github PR since it doesn't notify us automatically
```

### CODE REVIEW

File: lang/funcs/core/net/ip_port.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	simple.ModuleRegister(ModuleName, "ip_port", &types.FuncValue{
+		T: types.NewType("func(ip, port str) str"),
+		V: IPPort,
+	})
+}
+
+// IPPort returns the combind IP:(input[0]) and Port:(input[1]) arguments
+// as long as port input is within range 1-65536
+func IPPort(input []types.Value) (types.Value, error) {
+	ip := input[0].Str()
```
Comment (frebib) [inline lang/funcs/core/net/ip_port.go:39]: We should validate the IP address too: https://golang.org/pkg/net/#ParseIP

File: lang/funcs/core/net/ip_port.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+	port := input[1].Str()
+	portInt, err := strconv.Atoi(port)
+
+	if err != nil {
+		return &types.StrValue{V: ""}, fmt.Errorf("err converting str to int %v", err)
+	}
+	if portInt < 1 || portInt > 65536 {
+		return &types.StrValue{V: ""}, errors.New("port number must be between 1-65536")
```
Comment (frebib) [inline lang/funcs/core/net/ip_port.go:47]: nit: the wording here is slightly confusing given the logic. I would expect either `between -0-65536` (exclusive) or preferably `between 1-65535` (inclusive). The current wording is exclusive-inclusive which is probably not what the user would expect.

Side note, in many cases a port number of `0` can be valid as it's used to denote "any port". Perhaps we should allow that

File: lang/funcs/core/net/ip_port_test.go [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+		name     string
+		ip       string
+		port     string
+		expected string
+		err      error
+	}{
+		// Success
+		{"success", "192.168.1.1", "80", "192.168.1.1:80", nil},
+		// Fail
+		{"fail", "192.168.1.1", "9483670", "", errors.New("port number must be between 1-65536")},
```
Comment (frebib) [inline lang/funcs/core/net/ip_port_test.go:38]: Can we get some IPv6 tests too?
Also some tests for invalid IP address/port combos if you add IP address parsing

**Review by frebib — CHANGES_REQUESTED** [review]

File: lang/funcs/core/net/ip_port.go (hunk 1 above)
Comment (S-ign) [reply lang/funcs/core/net/ip_port.go:39]: Oh ok, cool, didn't realize there was already an IP Parser, I'll add that! :)

File: lang/funcs/core/net/ip_port.go (hunk 2 above)
Comment (S-ign) [reply lang/funcs/core/net/ip_port.go]: Thanks, I updated this, hopefully, the wording is a little better.

File: lang/funcs/core/net/ip_port_test.go (hunk 3 above)
Comment (S-ign) [reply lang/funcs/core/net/ip_port_test.go]: net/ParseIP made this really easy to implement, thanks!

**Comment by purpleidea** [comment]
Side note, re discussion in IRC:

I noticed that one of the tests failed in an unusual place in this PR... And I said, gah, I hope it's not a race. So I dug into it, and it turns out it was caused by a bug in your code! (yay, no our fault!) So the good news is the tests work, the bad news is I think we could have caught this issue in a more obvious place. So I added a new commit that should help new developers out by providing a clearer error message.

https://github.com/purpleidea/mgmt/commit/887d374c530915565b0257e6ac6f8c0cda232067

You'll want to rebase your branch against this commit which is now in git master.

Cheers!

**Comment by purpleidea** [comment]
@S-ign Hey-- I'm not going to do a full review yet-- I'd like you to first make a small examples/lang/whatever.mcl that uses this function, and confirm you got it doing what you expect. This should replace the copied example that you committed in this PR.

When you've succeeded, please update this PR and put a comment here with the command you used to run this example =D

Thanks!

**Comment by purpleidea** [comment]
(For anyone else reading, @S-ign came to us for a learning experience, so this is not the traditional review process.)

**Pull request closed without merging**